src/sentry/tasks/on_demand_metrics.py

```diff
@@ -0,0 +1,273 @@
+            metrics.incr(
+                "task.on_demand_metrics.widget_queries.per_run.flag_disabled",
+                sample_rate=1.0,
+            )
+            continue
+
+        is_low_cardinality = _set_widget_query_low_cardinality(query, organization)
+        if not is_low_cardinality:
+            widget_query_high_cardinality_count += 1
+
+    metrics.incr(
+        "task.on_demand_metrics.widget_queries.per_run.high_cardinality",
+        amount=widget_query_high_cardinality_count,
+        sample_rate=1.0,
+    )
+    metrics.incr(
+        "task.on_demand_metrics.widget_queries.per_run.total",
+        amount=widget_query_count,
+        sample_rate=1.0,
+    )
+
+
+def _set_widget_query_low_cardinality(
+    widget_query: DashboardWidgetQuery, organization: Organization
+):
+    """
+    Checks cardinality of existing widget queries before allowing the metric spec, so that
+    group by clauses with high-cardinality tags are not added to the on_demand metric.
+
+    New queries will be checked upon creation and not allowed at that time.
+    """
+
+    max_cardinality_allowed = options.get("on_demand.max_widget_cardinality.count")
+    cache_key = _get_widget_query_cardinality_cache_key(widget_query)
+
+    # We default low cardinality to true since if it's false we'll remove user data.
+    is_low_cardinality = cache.get(cache_key, default=True)
+
+    with sentry_sdk.push_scope() as scope:
+        scope.set_tag("widget_query.widget_id", widget_query.id)
+        scope.set_tag("widget_query.org_id", organization.id)
+        scope.set_tag("widget_query.conditions", widget_query.conditions)
+
+        try:
+            processed_results, columns_to_check = _query_cardinality(widget_query, organization)
+            for column in columns_to_check:
+                count = processed_results["data"][0][f"count_unique({column})"]
+                if count > max_cardinality_allowed:
+                    cache.set(cache_key, False, timeout=_WIDGET_QUERY_CARDINALITY_TTL)
+                    scope.set_tag("widget_query.column_name", column)
+                    raise HighCardinalityWidgetException(
+                        f"Cardinality exceeded for dashboard_widget_query:{widget_query.id} with count:{count} and column:{column}"
+                    )
+            # If it's made it here then cardinality is low.
+            is_low_cardinality = True
+
+        except HighCardinalityWidgetException as error:
+            sentry_sdk.capture_exception(error)
+            is_low_cardinality = False
+        except SoftTimeLimitExceeded as error:
+            scope.set_tag("widget_soft_deadline", True)
+            sentry_sdk.capture_exception(error)
+        except Exception as error:
+            sentry_sdk.capture_exception(error)
+
+    _set_cardinality_cache(cache_key, is_low_cardinality)
+    return is_low_cardinality
+
+
+def _query_cardinality(widget_query: DashboardWidgetQuery, organization: Organization):
```

Comment:
My codecov extension tells me this code needs coverage:
<img width="72" alt="image" src="https://github.com/getsentry/sentry/assets/44410/5a1b5176-6d65-41f0-85f6-104b2c6272dc">

Reply:
Yep, I need to add a few more tests, was just splitting out the tests to not have a bunch of discover queries in the `pytest.parameterize`